include/psa/crypto.h

```diff
@@ -4009,6 +3974,545 @@ psa_status_t psa_generate_key(const psa_key_attributes_t *attributes,
+ * \param password              Identifier of the key holding the password or a
+ *                              value derived from the password (eg. by a
+ *                              memory-hard function).  It must remain valid
+ *                              until the operation terminates. It must be of
+ *                              type #PSA_KEY_TYPE_PASSWORD or
+ *                              #PSA_KEY_TYPE_DERIVE. It has to allow the usage
+ *                              #PSA_KEY_USAGE_DERIVE.
+ *
+ * \retval #PSA_SUCCESS
+ *         Success.
+ * \retval #PSA_ERROR_CORRUPTION_DETECTED
+ * \retval #PSA_ERROR_INVALID_HANDLE
+ * \retval #PSA_ERROR_COMMUNICATION_FAILURE
+ * \retval #PSA_ERROR_HARDWARE_FAILURE
+ * \retval #PSA_ERROR_STORAGE_FAILURE
+ * \retval #PSA_ERROR_NOT_PERMITTED
+ * \retval #PSA_ERROR_INVALID_ARGUMENT
+ *         \p key is not compatible with the algorithm in \p cipher_suite.
+ * \retval #PSA_ERROR_BAD_STATE
+ *         The library has not been previously initialized by psa_crypto_init().
+ *         It is implementation-dependent whether a failure to initialize
+ *         results in this error code.
+ */
+psa_status_t psa_pake_set_password_key(psa_pake_operation_t *operation,
+                                       mbedtls_svc_key_id_t password);
+
+/** Set the password for a password-authenticated key exchange with memory hard
+ * function.
+ *
+ * Some protocols require using values derived from passwords via memory hard
+ * functions to mitigate dictionary attacks. Memory hard functions can be
+ * accessed through the key derivation interface and the result can be supplied
+ * to the PAKE operation in the form of a key derivation object.
+ *
+ * \param[in,out] operation     The operation object to set the password for. It
+ *                              must have been set up by psa_pake_setup() and
+ *                              not yet in use.
+ * \param key_derivation        An ongoing key derivation operation set up from
+ *                              the password and in a state suitable for
+ *                              calling psa_key_derivation_output_bytes().
+ *
+ * \retval #PSA_SUCCESS
+ *         Success.
+ * \retval #PSA_ERROR_CORRUPTION_DETECTED
+ * \retval #PSA_ERROR_INVALID_HANDLE
+ * \retval #PSA_ERROR_COMMUNICATION_FAILURE
+ * \retval #PSA_ERROR_HARDWARE_FAILURE
+ * \retval #PSA_ERROR_STORAGE_FAILURE
+ * \retval #PSA_ERROR_NOT_PERMITTED
+ * \retval #PSA_ERROR_INVALID_ARGUMENT
+ *         \p key_derivation is not ready for a call to
+ *         psa_key_derivation_output_bytes().
+ * \retval #PSA_ERROR_BAD_STATE
+ *         The library has not been previously initialized by psa_crypto_init().
+ *         It is implementation-dependent whether a failure to initialize
+ *         results in this error code.
+ */
+psa_status_t psa_pake_set_password_mhf(psa_pake_operation_t *operation,
+                                       psa_pake_operation_t *key_derivation);
```

Comment:
Since both of these are operation objects technically both are [in,out] parameters. `operation` takes an input from `key_derivation`, but its state changes as well. The same thing happens for `psa_pake_get_implicit_key()`, just the way around.

From the perspective of thinking in objects both of these functions are better thought of as related to the PAKE operation, because they need direct/low level access to the PAKE structure, whereas (in theory) they can use the public API of the key derivation operation without any need for low level access to the key derivation structure.

I put the PAKE operation in both parameter lists first, because that is the object that these functions belong to. And by "belong to" I mean that in other languages this wouldn't need to be passed and would be referred to as `self` or `this`. In cases like this I think it makes sense to put the "self/this" structure first in the parameter list.